docs/design/proposals/nvme-of.md

```diff
@@ -0,0 +1,249 @@
+volumes that can be used over NVMe-oF TCP. This makes it possible for Windows
+container hosts to consume storage from a Ceph cluster (RBD). It also allows
+external hosts to use a standard protocol (not Ceph RBD) to attach
+block-devices.
+
+The below diagram shows how the general architecture of the involved components
+will look like:
+
+```
+           .--------------------.
+           | Container Platform |
+           |     Kubernetes     |
+           '--------------------'
+              ^             ^
+              |             |
+              v             v
+.-------------------.    .------------------.
+| Controller-plugin |    |    Node-plugin   |
+|   Provisioner     |    | Attacher/Mounter |
+'-------------------'    '------------------'
+   ^          ^             ^
+   |          |             |
+   |          v             v
+   |        .-----------------.
+   |        | NVMe-oF Gateway |
+   |        |   ceph-nvmeof   |
+   |        '-----------------'
+   |          ^
+   |          |
+   v          v
+ .--------------.
+ | Ceph cluster |
+ '--------------'
+```
+
+:small_blue_diamond:Container Platform: calls CSI procedures to administrate
+the storage.
+
+:small_blue_diamond:Controller-plugin: Ceph-CSI instance handling the CSI
+Controller procedures.
+
+:small_blue_diamond:Node-plugin: Ceph-CSI instance handling the CSI Node
+procedures.
+
+:small_blue_diamond:NVMe-oF Gateway: one or more instances if `ceph-nvmeof`.
+
+:small_blue_diamond:Ceph cluster: Storage backend with RBD.
+
+## :wrench: Requirements
+
+The Minimal Viable Product should provide the most important functionalities
+that users expect. There is a need to be prepared for extending the
+functionality in the future.
+
+### CSI Controller-plugin / Provisioner
+
+The Controller-plugin should implement the following CSI procedures:
+
+- `ControllerCreateVolume` / `ControllerDeleteVolume`
+  1. Create RBD-image on the Ceph cluster (see Considerations)
+  1. Create the NVMe-subsystem (see Considerations)
+  1. Expose the RBD-image through the NVMe-oF Gateway
+
+- `ControllerPublishVolume` / `ControllerUnpublishVolume`
+  1. Allow access to the subsystem from the host where the Node-plugin will
+     attach the volume
+
+There are several parameters that can be configurable when a user creates a
+volume. Some of these parameters are set by the Container Platform
+administrator, others may be configurable by users.
+
+Administrator settable parameters (set in a StorageClass):
+
+- Ceph cluster
+- Credentials (Ceph user/key)
+- Ceph pool
+- RBD-image features
+- NVMe-oF Gateway (or gateway group?)
+- NVMe subsystem
+
+User controlled parameters (set in a PersistentVolumeClaim):
+
+- size of the volume
+- filesystem or raw-block access
+
+### CSI Node-plugin / Attacher & Mounter
+
+The Node-plugin should implement the following CSI procedures:
+
+- `NodeStageVolume` / `NodeUnstageVolume`
+  1. attach the NVMe-oF namespace from the NVMe-oF Gateway
```

Comment:
Do we have any limit on the nodes here?

Reply:
What limits are you thinking about? I am not aware of anything specific at the moment.

Reply:
i dont know, like we have limit for number of snapshots on an image for mapping the node. If we dont have any limits in gateway itself we are good